### PR TITLE
feat(performance): verify package artifact relocations

### PR DESCRIPTION
**Pull request by paulfalgout**
## Linked issue

- No linked issue; this is preparatory roadmap work for the adapter package extraction.

## What

- Discover and measure runtime artifacts and production graphs from the planned `@marionette/adapters` package alongside core.
- Add a canonical one-to-one relocation contract that preserves exact-base artifact and graph provenance.
- Revalidate relocation-bearing approval reports against exact-base and candidate contracts before rendering them.

## Agent-development failure addressed

- Exact-base performance checks previously treated a moved artifact as either a deletion or an unrelated new artifact, so the adapter extraction could not preserve measured provenance.
- The first implementation also allowed report rendering to consume relocation metadata without independently binding it to the candidate contract; this PR now fails closed on missing, invalid, or tampered contract evidence.

## Public behavior contract

- Canonical behavior after this PR: current exports and runtime behavior are unchanged; performance governance can verify an explicitly declared package relocation.
- Superseded behavior removed, if any: none for current consumers.
- Compatibility exception and removal condition, if any: none. Merged relocation records become immutable provenance.

## Runtime cost

- [ ] Static or documentation only
- [x] Development or test only; excluded from production entrypoints
- [ ] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: CI measured 0 B change for every current runtime artifact; aggregate remains 90,107 / 100,000 Brotli bytes.
- Startup/hot path: no production code changes; hosted timing remains reporting-only.
- Allocations/retention: deterministic resource comparison passed with no increase.

## Validation

- [x] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
npm run test:performance  # 118 passed
npm run lint:ci           # passed
git diff --check          # passed
```

Coverage includes repeated temporary-package/Rollup measurement, malformed and conflicting relocation maps, uncovered removals, immutable merged provenance, tampered report metadata, and the exact report CLI arguments used by CI.

## Package and browser impact

- Entry points or install fixtures affected: no current entry point changes; the verifier can discover the future `packages/adapters` manifest and Rollup config.
- Browser contracts affected: none.
- Production/dev/test module-graph impact: performance tooling only; no production module graph changed.

## Rollback or deprecation

- This commit can be reverted before a dependent relocation lands. Once the adapter extraction uses the contract, its relocation record must remain as immutable provenance.

## Deployment Impact

- No deployment or package publication is required. This is repository CI governance only.

## Review notes

- Review feedback led to independent report-time contract validation, fail-closed relocation cases, preserved external-import baselines, copy-on-read Rollup configuration handling, and repeated-measurement coverage.

### CODE REVIEW

**Comment by paulfalgout** [comment]
**Approved** — reviewed at fetched SHA `e45750f5` (base `8b755bf5`). No `dist/` change, so the core aggregate stays at **90,107 / 100,000**.

I reviewed this adversarially rather than descriptively, since the PR's central claim is that a candidate cannot "disguise deletions, collisions, or stale approval data as relocations." That claim holds. The specific attacks I tried to construct and could not:

**Deleting an artifact while declaring it moved.** `newProductionReportDelta` excuses a missing base artifact when it appears as a `from`, which on its own would let a candidate declare a move and simply never produce the target. The contract layer closes it with a four-way check:

```js
if (!authorityArtifacts.map.has(from) || candidateArtifacts.map.has(from) ||
    authorityArtifacts.map.has(to) || !candidateArtifacts.map.has(to)) {
```

`from` must exist in base **and** be gone from the candidate; `to` must not pre-exist in base **and** must exist in the candidate. I went looking specifically for a missing `!candidateArtifacts.map.has(to)` and it is there.

**Collapsing two artifacts into one.** `artifactTargets.has(move.to)` rejects duplicate targets, so A→X and B→X cannot quietly delete B.

**Laundering a baseline through a move.** This is the one I expected to find open, because it is the most valuable to an attacker and the least obvious:

```js
if (movedArtifact && (artifact.name !== movedArtifact.name ||
    artifact.baselineBrotliBytes !== movedArtifact.baselineBrotliBytes)) {
```

A relocated artifact must carry its exact-base name and baseline, while a genuinely new one must declare `baselineBrotliBytes === 0`. Without that pairing, "relocation" would be a general-purpose way to reset an artifact's Phase-0 baseline — move it, declare it new, start from zero. Both halves are needed and both are present.

**Decoupling a graph from its artifact.** Line 594 cross-links the two maps: `relocation.artifactMoves.get(movedGraph.output) !== graph.output`. A relocated graph must point at the relocated output, so the artifact and graph relocations cannot disagree.

**Reusing or editing a relocation later.** `relocationTransition` returns early with a violation when the authority contract already carries `relocations` and the candidate changes them. Relocations therefore apply exactly once, in the transition itself, and become immutable provenance afterward.

Canonical form is enforced throughout — exact object fields, sorted unique `from` values, no self-moves, `..`/`//` rejected in paths — which matters because it removes reordering and shape variation as places for ambiguity to hide.

### Notes, not objections

The `validArtifactPath` pattern now admits `packages/<name>/dist/**`, which is what makes the `@marionette/adapters` extraction measurable. Worth flagging that this is the moment the performance authority stopped being core-only; the `dist|packages/*/dist` alternation is now load-bearing for every future package.

Building this **before** the extraction is the right sequence. Landing the relocation contract in the same PR that moves the artifacts would mean the mechanism validating the move and the move itself arrive together, with no prior state in which the mechanism was exercised against a known-good tree — and, given that a relocation is self-describing, that is exactly the situation where a subtly permissive check would pass unnoticed.

This should also settle the consumer-fixture question I raised on #377 and #381: with `from`/`to` provenance available, the adapter extraction has a way to describe `marionette/backbone` and `marionette/jquery-dom-api` moving out of core without the fixture guard reading it as deletion.

I did not review the CI wiring in `.github/workflows/ci.yml` beyond noting it changed, or the `contract.test.mjs` cases individually.

File: scripts/performance/growth-approval.mjs [hunk 1]
```diff
@@ -505,13 +584,16 @@ export function validateCandidateGrowthContract(
         !validArtifactPath(graph.output)) {
       violations.push(`New production graph ${subpath} has an invalid additive contract shape`);
     }
-    if (!Array.isArray(graph.baselineModules) || graph.baselineModules.length ||
-        !Array.isArray(graph.baselineExternalImports) || graph.baselineExternalImports.length) {
+    if (!Array.isArray(graph.baselineModules) || !Array.isArray(graph.baselineExternalImports) ||
+        (!movedGraph && (graph.baselineModules.length || graph.baselineExternalImports.length))) {
       violations.push(`New production graph ${subpath} Phase 0 module baselines must be empty`);
     }
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply scripts/performance/growth-approval.mjs:590]: Fixed at amended head 2ba701a5. Relocated graphs now require `baselineExternalImports` to deep-equal the exact-base graph, with a focused regression case that rewrites the baseline and expects rejection.

**Comment by paulfalgout** [comment]
**Approval stands** at `2ba701a5`. Two of these are real bugs, and one is the same trust-boundary class as the `packageName` fix in #384.

### The imported config array was being mutated

```diff
- ({ default: configurations } = await import(configUrl.href));
+ const { default: rootConfigurations } = await import(configUrl.href);
+ configurations = [...rootConfigurations];
```

ES module namespaces are cached, so appending the adapters configuration was mutating the array owned by `rollup.config.mjs` itself. A second `measure()` in the same process would inherit the already-appended entry and push again. The new `repeated` assertion in the fixture test is the right way to pin it — the failure only exists on the second call, which is precisely why nothing caught it.

### The timing harness revision was self-attested

```diff
- { timingHarnessRevision: candidateContract.timing?.harnessRevision },
+ { timingHarnessRevision: Object.hasOwn(authorityContract, 'timing') ?
+     await committedTimingHarnessRevision() : undefined },
```

The candidate contract was supplying the value it was then validated against, so any declared revision validated trivially. Reading it from the committed harness instead makes the check mean something. This is the same shape as taking the package name from the in-repo allowlist rather than the manifest in #384 — the artifact under validation should never be the source of the expected value. Worth noting it only reached the relocation path introduced by this PR, so nothing shipped with it.

### The graph baseline pairing is now half complete

Adding the `baselineExternalImports` equality check for relocated graphs closes the gap I described for artifacts, applied to graphs. But the sibling field did not get the same treatment:

```js
// new graphs: both must be empty
(!movedGraph && (graph.baselineModules.length || graph.baselineExternalImports.length))

// relocated graphs: only external imports are compared
if (movedGraph && !isDeepStrictEqual(graph.baselineExternalImports, movedGraph.baselineExternalImports))
```

`baselineModules` is checked only for `Array.isArray` when a graph is relocated. So a relocation can still declare an arbitrary module baseline while preserving external imports exactly — which is the same laundering route you just closed, one field over. A relocated graph is supposed to be the *same* graph at a new subpath; if its external-import baseline must survive the move unchanged, its module baseline has no more reason to be editable.

The fix looks like one more `isDeepStrictEqual` beside the one just added. I would not hold the PR for it, but it is worth doing here rather than after the extraction lands, since afterwards the relocation becomes frozen provenance and the discrepancy is baked into the authority contract.

**Comment by paulfalgout** [comment]
**Approved at exact amended head `2ba701a57b7ebacd6381e4eda70733c1ae6d9bd8`** (base `8b755bf5d69564379a98aea993d1114a235c8f9e`).

The amended head addresses every actionable CodeRabbit finding: repeated measurements no longer mutate the cached Rollup configuration, relocated graphs preserve the exact-base external-import baseline, and report-time timing validation uses the committed harness revision. Focused regression coverage, `npm run test:performance` (118/118), and `npm run lint:ci` pass locally. Fresh CI is fully green, the only review thread is resolved, the PR is mergeable, and bundle-size reports 0 B change for all current artifacts.
